### PR TITLE
perf(ext/flash): vectored writes for sendfile headers

### PR DESCRIPTION
**Pull request by littledivy**
This felt like a good place to introduce the `UnixIoSlice` abstraction - which we will need for normal vectored writes in #15571

### CODE REVIEW

File: ext/flash/lib.rs [hunk 1]
```diff
@@ -244,12 +243,15 @@ async fn op_flash_write_resource(
         let mut stat: libc::stat = unsafe { std::mem::zeroed() };
         // SAFETY: call to libc::fstat.
         if unsafe { libc::fstat(fd, &mut stat) } >= 0 {
-          let _ = sock.write(
-            format!("Content-Length: {}\r\n\r\n", stat.st_size).as_bytes(),
-          );
+          let cnt_length = format!("Content-Length: {}\r\n\r\n", stat.st_size);
```
Comment (bartlomieju) [inline ext/flash/lib.rs:246]: This is an additional allocation, maybe we could create 3 `UnixIoSlice` instances containing:
- `Content-Length: `
- `{stat.st_size}`
- `\r\n\r\n`

Comment (littledivy) [reply ext/flash/lib.rs]: Fixed. Now uses `itoa` which uses a stack allocated buffer for formating st_size...we're doing zero heap allocation now.

File: ext/flash/sendfile.rs [hunk 2]
```diff
@@ -21,9 +83,50 @@ impl SendFile {
...
+          )
+        };
+
+        let io_vec = &mut self.slices;
+        let accumulated_len = advance_io_vec(io_vec, length as usize);
+        if io_vec.is_empty() {
+          self.sending_headers = false;
+          self.written += length as usize - accumulated_len;
+        }
+        return Ok(length as usize);
```
Comment (bnoordhuis) [inline ext/flash/sendfile.rs:111]: I don't quite understand this return statement. Don't you want to move on to the actual sendfile call now?

File: ext/flash/sendfile.rs [hunk 3]
```diff
@@ -35,6 +138,42 @@ impl SendFile {
...
+          libc::sendfile(
+            self.io.0,
+            self.io.1,
+            self.written as libc::off_t,
+            &mut length,
+            &mut hdtr,
+            0,
+          )
+        };
+        if res == -1 {
```
Comment (bnoordhuis) [inline ext/flash/sendfile.rs:163]: You should special-case EAGAIN and EINTR here:

- EAGAIN and length == 0, treat as regular EAGAIN
- EAGAIN and length > 0, some (possibly all) data was written

Ditto for EINTR.

File: ext/flash/sendfile.rs [hunk 4]
```diff
@@ -21,9 +83,50 @@ impl SendFile {
...
+        // SAFETY: call to libc
+        let length = unsafe {
+          libc::setsockopt(
+            self.io.1,
+            libc::SOL_SOCKET,
+            libc::TCP_CORK,
+            &opt as *const _ as _,
+            4,
+          );
+          libc::writev(
```
[the remark this answers is not in the document]
Comment (bnoordhuis) [reply ext/flash/sendfile.rs:98]: Come to think of it, you can avoid the setsockopt() system calls if you switch to sendmsg() with the MSG_MORE flag. The sendfile() call undoes the effect of MSG_MORE.

File: ext/flash/sendfile.rs [hunk 5]
```diff
@@ -21,9 +83,50 @@ impl SendFile {
...
+      if self.sending_headers {
+        let opt = 1;
+        // SAFETY: call to libc
+        let length = unsafe {
+          libc::setsockopt(
+            self.io.1,
+            libc::SOL_SOCKET,
+            libc::TCP_CORK,
+            &opt as *const _ as _,
+            4,
```
[the remark this answers is not in the document]
Comment (billywhizz) [reply ext/flash/sendfile.rs:96]: will we ever have nodelay on in a web server? agree we don't need the cork if nagle is always on - might be something we want to allow consumer of api to set though?

[the remark this answers is not in the document]
Comment (bnoordhuis) [reply ext/flash/sendfile.rs:96]: Maybe? Nginx enables nodelay by default on some platforms (although not linux.)

**Pull request closed without merging**